Added changelog and Barrier op priority for Consumer batch purging issue

pranavrth opened the pull request:
- Added Changelog for the Consumer batch APIs multithreading issue
- Updated Barrier op priority

pranavrth commented:
@edenhill - Shall I add more details (explaining the issue and all?) to the changelog related to the issue?

pranavrth commented:
@edenhill - Shall I add partition check as well while purging? I think we don't need that right now as we are discouraging using any other operation while using consumer batch API.

Test 0122 was failing when I tried adding this check.

edenhill commented:
> Shall I add more details (explaining the issue and all?) to the changelog related to the issue?

Yes please, add a new "Known issues" section of the 2.0.0 release right after Upgrade considerations.

> Shall I add partition check as well while purging? I think we don't need that right now as we are discouraging using any other operation while using consumer batch API.

I think we really should add the partition check, since it will currently remove messages for other partitions.

pranavrth commented:
> Yes please, add a new "Known issues" section of the 2.0.0 release right after Upgrade considerations.

I had already added this section and added the issue detail. But I was asking whether I should go into the technical details of each of the problem? Can you please check once what I have added right now?

edenhill commented on CHANGELOG.md:
```diff
@@ -42,6 +42,20 @@ impersonation attacks) by default.
+
+ * The Consumer Batch APIs `rd_kafka_consume_batch` and `rd_kafka_consume_batch_queue`
+   are not thread safe if `rkmessages_size` is greater than 1 and any of the **seek**,
+   **pause**, **resume** and **repartition** operation is performed in parallel. Some
```
rebalancing, not repartition

edenhill commented on CHANGELOG.md:
```diff
@@ -42,6 +42,20 @@ impersonation attacks) by default.
+ * The Consumer Batch APIs `rd_kafka_consume_batch` and `rd_kafka_consume_batch_queue`
+   are not thread safe if `rkmessages_size` is greater than 1 and any of the **seek**,
+   **pause**, **resume** and **repartition** operation is performed in parallel. Some
+   of the messages might be lost in the above scenario.
```
might be lost, or erroneously returned to the application, in the above scenario.

edenhill commented on CHANGELOG.md:
```diff
@@ -42,6 +42,20 @@ impersonation attacks) by default.
+   It is strongly recommended to use the Consumer Batch APIs and the mentioned
+   operations in sequential order in order to get consistent result.
+   
+   For **repartition** operation to work in sequencial manner, please set `rebalance_cb`
```
rebalancing

edenhill commented on CHANGELOG.md:
```diff
@@ -42,6 +42,20 @@ impersonation attacks) by default.
+
+### Consumer issues
+
+ * The Consumer Batch APIs `rd_kafka_consume_batch` and `rd_kafka_consume_batch_queue`
```
Remove the bullet (*) and rename the "Consumer issues" chapter to "Poor Consumer batch API messaging guarantees"

edenhill commented on CHANGELOG.md:
```diff
@@ -42,6 +42,20 @@ impersonation attacks) by default.
+   For **repartition** operation to work in sequencial manner, please set `rebalance_cb`
+   configuration property (refer [examples/rdkafka_complex_consumer_example.c](examples/rdkafka_complex_consumer_example.c) for the help with the usage) for the consumer.
 
 ## Enhancements
```
To consumer fixes below, add what's been fixed.

edenhill commented on tests/0137-barrier_batch_consume.c:
```diff
@@ -0,0 +1,290 @@
+        TIMING_STOP(&t_cons);
+
+        for (i = 0; i < msg_cnt; i++) {
+            TEST_SAY("consumed message for partition %d with offset %lld\n", rkmessage[i]->partition, rkmessage[i]->offset);
```
Per message debug prints are typically too noisy and not that useful, use TEST_SAYL(3, "...") to only have them printed if you set TEST_LEVEL=3 (or higher), the default level is 2.

edenhill replied:
In fact, you already get this log automatically from test_msgver_add_msg() below for TEST_LEVEL=3, so remove this  for loop altogether.

pranavrth replied:
My bad. I printed those for the testing.

edenhill commented on tests/0137-barrier_batch_consume.c:
```diff
@@ -0,0 +1,290 @@
+                if (test_msgver_add_msg(rk, arguments->mv, rkmessage[i]) == 0)
+                        TEST_FAIL(
+                            "The message is not from testid "
+                            "%" PRId64 " \n",
```
TEST_FAIL does not need a newline, it is automatically added.
TEST_SAY does though.

edenhill commented on tests/0137-barrier_batch_consume.c:
```diff
@@ -0,0 +1,290 @@
+}
+
+
+static void do_test_consume_batch_with_seek() {
```
```suggestion
static void do_test_consume_batch_with_seek(void) {
```

C: functions without parameters: accepts any parameters (never desired)
C++: functions without parameters: no parameters.

edenhill commented on tests/0137-barrier_batch_consume.c:
```diff
@@ -0,0 +1,290 @@
+            thrd_success)
+                TEST_FAIL("Failed to create thread for %s", "CONSUMER");
+
+        rkt = test_create_producer_topic(consumer, topic);
```
Even though it will work it is not right to use a producer function on a consumer (might break in the future).

I recommend you use rd_kafka_seek_partitions() instead which does not require a topic object, rd_kafka_seek() is being deprecated.

edenhill commented on tests/0137-barrier_batch_consume.c:
```diff
@@ -0,0 +1,290 @@
+        rkt = test_create_producer_topic(consumer, topic);
+        err = rd_kafka_seek(rkt, seek_partition, seek_offset, 3000);
+
+        if(err == RD_KAFKA_RESP_ERR__TIMED_OUT) {
```
Run `make style-fix`  (install clang-format-10 first)

edenhill replied:
```c
TEST_ASSERT(err == RD_KAFKA_RESP_ERR__TIMED_OUT,
   "Failed to seek.....");`
```

But I really think you should check for any error, not just this one.

edenhill commented on tests/0137-barrier_batch_consume.c:
```diff
@@ -0,0 +1,290 @@
+                           expected_msg_cnt);
+        test_msgver_clear(&mv);
+
+        rd_kafka_topic_destroy(rkt);
```
using these topic_t objects is generally not recommended anymore, it is typically a thing of the past and only really useful for high-performant producing.

edenhill commented on tests/0137-barrier_batch_consume.c:
```diff
@@ -0,0 +1,290 @@
+        test_msgver_t mv;
+        thrd_t thread_id;
+        rd_kafka_resp_err_t err;
+        rd_kafka_topic_partition_list_t *parition_list;
```
typo

edenhill commented on tests/0137-barrier_batch_consume.c:
```diff
@@ -0,0 +1,290 @@
+        parition_list = rd_kafka_topic_partition_list_new(1);
+        rd_kafka_topic_partition_list_add(parition_list, topic, pause_partition);
+
+        sleep(1);
```
not portable, use rd_sleep()

edenhill commented on tests/0137-barrier_batch_consume.c:
```diff
@@ -0,0 +1,290 @@
+        sleep(1);
+        err = rd_kafka_pause_partitions(consumer, parition_list);
+
+        if(err != RD_KAFKA_RESP_ERR_NO_ERROR) {
```
Same as previous comment

edenhill commented on tests/0137-barrier_batch_consume.c:
```diff
@@ -0,0 +1,290 @@
+
+        err = rd_kafka_resume_partitions(consumer, parition_list);
+
+        if(err != RD_KAFKA_RESP_ERR_NO_ERROR) {
```
TEST_ASSERT(!err, "...")

edenhill commented on tests/0137-barrier_batch_consume.c:
```diff
@@ -0,0 +1,290 @@
+        thrd_join(thread_id, NULL);
+
+
+        rd_kafka_message_t **abc =
```
No variable declarations after statements, all varaibles declarations must go at the top of the scope.

pranavrth replied:
I am coming from a java background where it is advised to use variable declaration as near the first usage of it as possible. For my understanding, can you please tell me why is it important to have variable declaration at the top of the scope?

edenhill replied:
Up until recently, the code wouldn't build on older MSVC compilers if variables were declared after statements, so that's why that's the current style in librdkafka.

edenhill commented on tests/0137-barrier_batch_consume.c:
```diff
@@ -0,0 +1,290 @@
+        rd_kafka_message_t **abc =
+            malloc(consume_msg_cnt * sizeof(*abc));
+
+        int msg_cnt = (int)rd_kafka_consume_batch_queue(rkq, timeout_ms, abc,
```
all variable decl at top of scope

edenhill replied:
don't cast here, use the proper type instead

edenhill commented on tests/0137-barrier_batch_consume.c:
```diff
@@ -0,0 +1,290 @@
+int main_0137_barrier_batch_consume(int argc, char **argv) {
+        do_test_consume_batch_with_seek();
+        // Run this test once consume batch is fully fixed.
+        // do_test_consume_batch_with_pause_and_resume();
```
:+1:

edenhill commented on tests/0137-barrier_batch_consume.c:
```diff
@@ -0,0 +1,290 @@
+
+int main_0137_barrier_batch_consume(int argc, char **argv) {
+        do_test_consume_batch_with_seek();
+        // Run this test once consume batch is fully fixed.
```
```suggestion
        // FIXME: Run this test once consume batch is fully fixed.
```

edenhill reviewed (changes requested):
Very good stuff! Just some minor comments

edenhill commented:
And make sure to run `make style-fix`

pranavrth commented:
Fixes - https://github.com/confluentinc/confluent-kafka-python/issues/1443

edenhill commented on CHANGELOG.md:
```diff
@@ -42,6 +42,21 @@ impersonation attacks) by default.
+
+The Consumer Batch APIs `rd_kafka_consume_batch()` and `rd_kafka_consume_batch_queue()`
+are not thread safe if `rkmessages_size` is greater than 1 and any of the **seek**,
+**pause**, **resume** and **rebalancing** operation is performed in parallel. Some
```
```suggestion
**pause**, **resume**, or **rebalancing** operation is performed in parallel. Some
```

edenhill reviewed (approved):
LGTM

edenhill commented:
For CIs to run you need to push this branch to confluentinc/librdkafka with the name `dev_pr4077`, can you please do that and open a new PR for that branch?

edenhill commented:
Also make sure you rebase your branch on latest master, there are conflicts on this one.

pranavrth commented:
Closing this as the changes are merged with PR - https://github.com/confluentinc/librdkafka/pull/4117
